www/apps/book/app/advanced-development/events-and-subscribers/data-payload/page.mdx

```diff
@@ -29,7 +30,7 @@ import type {
 export default async function productCreateHandler({
   data,
 }: SubscriberArgs<{ id: string }>) {
-  const productId = "data" in data ? data.data.id : data.id
+  const productId = data.data.id
```

Comment:
thought(non-blocking): Still not super happy about the `data.data`. Feels like something is not right. 

@adrien2p, do you have any suggestions here? Can we change the input to the subscriber to `body` instead, so that you would do:
```
export default async function productCreateHandler({
  body,
}: SubscriberArgs) {
  const { data, metadata } = body
  const productId = data.id
```

@shahednasser, please feel free to also pitch in.

Reply:
`body` sounds ambigious. we could instead keep `data` as the data payload and pass the `metadata` and `eventName` in a new parameter `context`

Reply:
The reason I suggested body is that it is the name of the [type](https://github.com/medusajs/medusa/blob/b3351da264f2da1db11ca2b9420566903de7545b/packages/medusa/src/types/subscribers.ts#L13), but we can change it.

It is worth noting `data` is not the event data payload. It is the entire message content. The actually event data is what we current have at `data.data`. 

But perhaps we should clean up the types here a bit to something along the lines of:
```
export type SubscriberArgs<T = unknown> = {
  body: MessageBody<T> // or something else like `content`
  container: MedusaContainer
  pluginOptions: Record<string, unknown>
}

export type MessageBody<TData = unknown> = {
  eventName: string
  metadata?: EventMetadata
  data: TData
}
``` 

I think having the `eventName` both in the subscriber arguments and the message content is also confusing. This type refactor would create a clear separation of the concerns. 

Let me know what you think. I am also open to other suggestions.

Reply:
What about calling it `event`? 

So you would do:
```
export default async function productCreateHandler({
  event,
}: SubscriberArgs) {
  const { data, metadata } = event
  const productId = data.id
```

Reply:
> It is worth noting data is not the event data payload. It is the entire message content. The actually event data is what we current have at data.data.

yes I'm aware, I was suggesting splitting it up so we have two properties instead of one `data` property: `data` and `context`.

But I like the `event` option too that you mentioned 👍🏻

Reply:
> yes I'm aware, I was suggesting splitting it up so we have two properties instead of one data property: data and context.

Ah I see. 

> But I like the event option too that you mentioned

Let me create a ticket for this and we can tackle it later. Don't want to block this PR 👍

Reply:
That would require us to change a lot of terminology for users though. The user knows that they're handling an emitting event, they don't know about the underlying work of a message being sent/received.

Reply:
yes thats true, should we rename data inside data to payload maybe?

Reply:
both work for me 👍🏻

Reply:
Just linking the [Linear issue](https://linear.app/medusajs/issue/TRI-5/revisit-arguments-passed-to-subscribers) for futher discussions.